### ISSUE TITLE
Printed space goes missing when using grep --color at end of line

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#3362]
Might be a combination of the space and the \x1b[0m?

<img width="825" alt="Screen Shot 2021-06-08 at 7 41 48 AM" src="https://user-images.githubusercontent.com/2193314/121206159-179a3e00-c82d-11eb-9768-d1483ad79891.png">

VS Code issue: https://github.com/microsoft/vscode/issues/125605

**Comment by Tyriar** [comment]
Space is getting written to 86, so is a:

<img width="177" alt="Screen Shot 2021-06-08 at 7 53 58 AM" src="https://user-images.githubusercontent.com/2193314/121208139-b1161f80-c82e-11eb-83ce-8868062c42ea.png">

**Comment by mofux** [comment]
I can reproduce as well, also only with `--color=always`. Resizing the terminal (reflow) doesn't fix or produce the bug, so I guess it's happening somewhere during parsing. Maybe some limiting logic (`Math.max`, `Math.min`, ...) that causes an off-by-one error for this edge case?

**Comment by jerch** [comment]
It is caused by this line: https://github.com/xtermjs/xterm.js/blob/19c9fef8e2579cb3c755325ea4807e7dc70b406b/src/common/InputHandler.ts#L1322

xterm.js inherits this behavior from xterm, disabling it here will let the xterm test `t0055-EL.in` fail.

**Comment by mofux** [comment]
@jerch In #2979 you conclude:

> If this turns out as a major limitation, we could try to lift the cursor restriction up to term.cols instead of term.cols - 1 for EL (prolly ED as well). But this needs serious testing to not introduce side effects on other ends.

Do you think that would fix this particular bug?

I've just run the test case above in `Terminal.app` and it does not show this bug.

**Comment by jerch** [comment]
Yes I think it will fix both of these issues, maybe all wee need is this:
```typescript
  this._restrictCursor(this._bufferService.cols);
```
which would allow the cursor to stick up to cols. Technically this is an error for any cursor sequences, since the cell is undefined. But we use the same trick during normal `print` to get the wrapping done, therefore I think it should not cause further trouble. EL internal code itself is guarded against overflows.

**Comment by Tyriar** [comment]
Silly me just debugged into it to find the cause before I read the responses 😛 

@jerch I'll defer to you for how we should handle this, my thinking is we should align with the modern terminals for this as it feels like a bug. Any testing in particular you had in mind? If iTerm/VTE behave like this it's probably a fairly safe change?

**Comment by jerch** [comment]
@Tyriar Yes it feels like a bug to me. Also #2979 indicates, that most TEs allow the cursor overflow here (not resetting wrap-around). Imho we should fix it for ED/EL.

Regarding tests - the renderers should not overflow here, as we do the same with print just fine. But I need to check other cursor sequences, if they restrict the cursor in a pre- or post step and if their internal code is properly guarded (some test cases should do).
